exec/supervisor/plugin.go

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2019 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package supervisor
+
+import (
+	"os/exec"
+	"sync"
+
+	pm "github.com/ligato/cn-infra/exec/processmanager"
+	"github.com/ligato/cn-infra/exec/processmanager/status"
+	"github.com/ligato/cn-infra/infra"
+	"github.com/pkg/errors"
+)
+
+const unnamedProcess = "<unnamed>"
+
+// Supervisor defines methods to obtain information about running instances
+type Supervisor interface {
+	// GetProcessNames returns names of all running process instances
+	GetProcessNames() []string
+	// GetProcessByName returns an instance of given process
+	GetProcessByName(name string) pm.ProcessInstance
+}
+
+// Plugin is a supervisor plugin structure
+type Plugin struct {
+	// a list of active process instances
+	mx        sync.Mutex
+	instances map[string]pm.ProcessInstance
+
+	config *Config
+	Deps
+}
+
+// Deps are supervisor dependencies (standard infra plugins + process manager)
+type Deps struct {
+	pm pm.ProcessManager
+	infra.PluginDeps
+}
+
+// Init starts processes and their watchers defined by the configuration file
+func (p *Plugin) Init() error {
+	p.instances = make(map[string]pm.ProcessInstance)
+
+	if err := p.getConfig(); err != nil {
+		return err
+	}
+
+	// run watchers even if there are failed processes. They will be handled
+	// as terminated and exit immediately
+	err := p.start(p.config.Processes)
+	p.watch()
+
+	return err
+}
+
+// Close stops all remaining processes
+func (p *Plugin) Close() error {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	// TODO option to keep them running?
+
+	for name, instance := range p.instances {
+		if _, err := instance.StopAndWait(); err != nil {
+			p.Log.Errorf("supervisor close error: failed to stop process %s: %v", name, err)
+		}
+	}
+
+	return nil
+}
+
+// GetProcessNames returns names of all running process instances
+func (p *Plugin) GetProcessNames() (names []string) {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	for name := range p.instances {
+		names = append(names, name)
+	}
+	return names
+}
+
+// GetProcessByName returns an instance of given process
+func (p *Plugin) GetProcessByName(reqName string) pm.ProcessInstance {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	for name, instance := range p.instances {
+		if name == reqName {
+			return instance
+		}
+	}
+	return nil
+}
+
+func (p *Plugin) start(processes []Process) error {
+	pmLogManager := newPmLoggerManager()
```

Comment:
The entire logging for programs is totally over-engineered. Adding to my [previous review comment](https://github.com/ligato/cn-infra/pull/393/#discussion_r283307806):

https://github.com/ligato/cn-infra/blob/8933a15080a07ab92ae50bd8ccab73da5455f97f/exec/supervisor/pmlogger.go#L57

There is no need for logs to go through logrus, the [`log.Logger`](https://golang.org/pkg/log/#Logger) from stdlib is enough for our use-case, even that seems unnecessary when logging to actual file. 

https://github.com/ligato/cn-infra/blob/8933a15080a07ab92ae50bd8ccab73da5455f97f/exec/supervisor/pmlogger.go#L110

This variable is useless. Just use the log from Plugin.

https://github.com/ligato/cn-infra/blob/8933a15080a07ab92ae50bd8ccab73da5455f97f/exec/supervisor/pmlogger.go#L38

There is no need for channels. It's okay to open file more than once when using it in multiple programs.

https://github.com/ligato/cn-infra/blob/8933a15080a07ab92ae50bd8ccab73da5455f97f/exec/supervisor/pmlogger.go#L42

This is not even called and even if it was, it might panic because it's not closed by [the producer](https://github.com/ligato/cn-infra/blob/8933a15080a07ab92ae50bd8ccab73da5455f97f/exec/supervisor/pmlogger.go#L124).

---

And even with all of this code for logging it does not support the simple use-case which we currently have in vpp-agent, which is: if log file is not defined, use stdout/stderr or program that runs supervisor. We want to keep this behaviour and we actually want to ensure that the output printed by child processes (agent in our case) is printed with color. This should return `true` in the agent process when using supervisor.